### ISSUE TITLE
chat server

### ISSUE THREAD

**Issue by Swananda-Punde** [rhklinux/game_framework#9]
I have edited progaram using std::list,
but still therer are errors regarding chat_client_online class.

Compile command : 
> g++ server.cpp chat_client_online.cpp -lboost_system -lpthread -o server

**chat_clinet_online.cpp**
```
#include<boost/asio.hpp>
#include<stdlib.h>
#include<iostream>

using std::string;
using boost::asio::ip::udp;

class chat_client_online {
	private :
		string client_name, ip;
		udp::endpoint endpoint;

		void set_name() {
			client_name = std::to_string(rand());
		}
	public :
		chat_client_online():client_name(""),ip(""),endpoint() {
			
		}
		chat_client_online(udp::endpoint &ep):endpoint(ep) {
			ip = endpoint.address().to_string();
			set_name();
		}
		string get_client_name() {
			return client_name;
		}
		udp::endpoint get_endpoint() {
			return endpoint;
		}
		~chat_client_online(){
			ip.clear();
			client_name.clear();
			//delete endpoint;
			delete this;
		}
};
```

**chat_clinet_online.h**
```
#ifndef chat_client_online
#define chat_client_online

class chat_client_online;

#endif
```

**server.cpp**
```
#include<iostream>
#include<string>
#include<boost/asio.hpp>
#include<list>
#include<iterator>
#include"chat_client_online.h"


using std::cout;
using std::cin;
using std::string;
using std::endl;
using boost::asio::ip::udp;
using std::list;

class server {

	private : 
		boost::asio::io_service io_service;
		list<chat_client_online> client_list;
		//chat_client_online *cc;

		bool add_client(chat_client_online &cc) {
			
			list<chat_client_online> :: iterator itr;
			for(itr = client_list.begin(); itr != client_list.end(); itr++)
			{
				if(((*itr)->data) == cc)
					return false;
			}
			client_list.push_front(cc);
			return true;
				
		}	
		void handle_send(udp::endpoint &client_endpoint, char *buffer)
		{
			chat_client_online cc(client_endpoint);
			//cc = new chat_client_online(client_endpoint);
			add_client(cc);
			list<chat_client_online> :: iterator itr;
			for(itr = client_list.begin(); itr != client_list.end(); itr++)
			{
				if((*itr)->data != cc)
				{
					udp::endpoint send_endpoint = ((*itr)->data).endpoint;
					socket.send_to(boost::asio::buffer(buffer, 512), send_endpoint);
				}
			}
		}
	public :
		server(boost::asio::io_service &io_service, int port)
		{
			char buffer[512];
			udp::socket socket(io_service, udp::endpoint(udp::v4(), port));
			udp::endpoint client_endpoint;
			size_t length = socket.receive_from(boost::asio::buffer(buffer,512),client_endpoint);

			cout << "\nReceved from Client : "<< buffer << endl;
			handle_send(client_endpoint, buffer);
			cout <<"\nIp associated : "<<client_endpoint.address().to_string()<<" end"<<endl;
		}

};//server class ends

int main(int argc, char *argv[])
{
	if(argc < 2 )
	{
		cout << "required arguments : ./server <port>"<<endl;
		return 0;
	}
	boost::asio::io_service io_service;
	server server(io_service, atoi(argv[1]));
}


```

**Comment by rhklinux** [comment]
Please post the errors.

```
#ifndef chat_client_online
#define chat_client_online

class chat_client_online;

#endif
```
One thing I could spot here is that this macro replaces class link_list with 'blank'.

**Comment by rhklinux** [comment]
Why you are implementing linked list ?
we already have std::list and std::vector in STL

**Comment by Swananda-Punde** [comment]
Errors : 
```
[swananda@swananda-pc server]$ g++ server.cpp chat_client_online.cpp -lboost_system -lpthread -o server
In file included from server.cpp:6:
chat_client_online.h:4:25: error: expected identifier before ‘;’ token
 class chat_client_online;
                         ^
server.cpp:20:26: error: wrong number of template arguments (0, should be at least 1)
   list<chat_client_online> client_list;
                          ^
In file included from /usr/include/c++/8.1.1/list:63,
                 from server.cpp:4:
/usr/include/c++/8.1.1/bits/stl_list.h:563:11: note: provided for ‘template<class _Tp, class _Alloc> class std::__cxx11::list’
     class list : protected _List_base<_Tp, _Alloc>
           ^~~~
server.cpp:23:38: error: expected identifier before ‘&’ token
   bool add_client(chat_client_online &cc) {
                                      ^
server.cpp: In member function ‘bool server::add_client(int&)’:
server.cpp:25:27: error: wrong number of template arguments (0, should be at least 1)
    list<chat_client_online> :: iterator itr;
                           ^
In file included from /usr/include/c++/8.1.1/list:63,
                 from server.cpp:4:
/usr/include/c++/8.1.1/bits/stl_list.h:563:11: note: provided for ‘template<class _Tp, class _Alloc> class std::__cxx11::list’
     class list : protected _List_base<_Tp, _Alloc>
           ^~~~
server.cpp:25:41: error: expected initializer before ‘itr’
    list<chat_client_online> :: iterator itr;
                                         ^~~
server.cpp:26:8: error: ‘itr’ was not declared in this scope
    for(itr = client_list.begin(); itr != client_list.end(); itr++)
        ^~~
server.cpp:26:26: error: request for member ‘begin’ in ‘((server*)this)->server::client_list’, which is of non-class type ‘int’
    for(itr = client_list.begin(); itr != client_list.end(); itr++)
                          ^~~~~
server.cpp:26:54: error: request for member ‘end’ in ‘((server*)this)->server::client_list’, which is of non-class type ‘int’
    for(itr = client_list.begin(); itr != client_list.end(); itr++)
                                                      ^~~
server.cpp:31:16: error: request for member ‘push_front’ in ‘((server*)this)->server::client_list’, which is of non-class type ‘int’
    client_list.push_front(cc);
                ^~~~~~~~~~
server.cpp: In member function ‘void server::handle_send(boost::asio::ip::udp::endpoint&, char*)’:
server.cpp:37:23: error: ‘cc’ was not declared in this scope
    chat_client_online cc(client_endpoint);
                       ^~
server.cpp:37:23: note: suggested alternative: ‘cc_t’
    chat_client_online cc(client_endpoint);
                       ^~
                       cc_t
server.cpp:40:27: error: wrong number of template arguments (0, should be at least 1)
    list<chat_client_online> :: iterator itr;
                           ^
In file included from /usr/include/c++/8.1.1/list:63,
                 from server.cpp:4:
/usr/include/c++/8.1.1/bits/stl_list.h:563:11: note: provided for ‘template<class _Tp, class _Alloc> class std::__cxx11::list’
     class list : protected _List_base<_Tp, _Alloc>
           ^~~~
server.cpp:40:41: error: expected initializer before ‘itr’
    list<chat_client_online> :: iterator itr;
                                         ^~~
server.cpp:41:8: error: ‘itr’ was not declared in this scope
    for(itr = client_list.begin(); itr != client_list.end(); itr++)
        ^~~
server.cpp:41:26: error: request for member ‘begin’ in ‘((server*)this)->server::client_list’, which is of non-class type ‘int’
    for(itr = client_list.begin(); itr != client_list.end(); itr++)
                          ^~~~~
server.cpp:41:54: error: request for member ‘end’ in ‘((server*)this)->server::client_list’, which is of non-class type ‘int’
    for(itr = client_list.begin(); itr != client_list.end(); itr++)
                                                      ^~~
server.cpp:46:13: error: request for member ‘send_to’ in ‘socket’, which is of non-class type ‘int(int, int, int) throw ()’ {aka ‘int(int, int, int)’}
      socket.send_to(boost::asio::buffer(buffer, 512), send_endpoint);
```

**Comment by Swananda-Punde** [comment]
I have solved these errors.
the new header file is
**chat_client_online.h**
```
#ifndef CHAT_CLIENT_ONLINE//chat_client_online
#define CHAT_CLIENT_ONLINE//chat_client_online

#include<boost/asio.hpp>
#include<stdlib.h>
#include<iostream>

using std::string;
using boost::asio::ip::udp;

class chat_client_online{
	private :
		string client_name, ip;
		udp::endpoint endpoint;

		void set_name();
	public :
		chat_client_online();
		chat_client_online(udp::endpoint&);
		string get_client_name();
		udp::endpoint get_endpoint();
		bool operator== (chat_client_online &cc) {
			if(cc.endpoint == this->endpoint)
				return true;
			return false;
		}
		~chat_client_online();
};

#endif
```
